### PR TITLE
Fix email validation to enforce valid TLD

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb This is ready for review.

File: core/lib/spree/core/validators/email.rb [hunk 1]
```diff
@@ -2,7 +2,20 @@ class EmailValidator < ActiveModel::EachValidator
   LOCAL_PART_MAX_LENGTH = 64
   HOSTNAME_MAX_LENGTH   = 255
   LOCAL_PART_PATTERN    = /\A[^@\s\.](?:\.?[^@\s\.])*\z/i.freeze
-  HOSTNAME_PATTERN      = /\A(?:[-\p{L}\d]{1,63}\.)+\p{L}{2,}\z/i.freeze
+
+  TLDS =
```
Comment (dkubb) [inline core/lib/spree/core/validators/email.rb]: I'd recommend adding a comment that includes a URL for where this domain list can be downloaded.

File: core/lib/spree/core/validators/email.rb [hunk 2]
```diff
@@ -2,7 +2,20 @@ class EmailValidator < ActiveModel::EachValidator
...
+    IceNine.deep_freeze(
+      Spree::ROOT
+      .join(*%w[config tlds-alpha-by-domain.txt])
+      .read
+      .split
+      .reject { |line| line.start_with?('#') }
```
Comment (dkubb) [inline core/lib/spree/core/validators/email.rb]: You could change this to:

```ruby
.grep(/\A[^#]/)
```

Comment (mbj) [inline core/lib/spree/core/validators/email.rb]: I was considering `String#start_with?` more primitve than a regexp, but the regexp has no block, and so in sum its less complex, I'll switch to it.

**Comment by mbj** [comment]
@dkubb Ready for the next pass, note I extended the scope to also reject quoted localparts.

File: core/lib/spree/core/validators/email.rb [hunk 3]
```diff
@@ -1,8 +1,22 @@
 class EmailValidator < ActiveModel::EachValidator
   LOCAL_PART_MAX_LENGTH = 64
   HOSTNAME_MAX_LENGTH   = 255
-  LOCAL_PART_PATTERN    = /\A[^@\s\.](?:\.?[^@\s\.])*\z/i.freeze
-  HOSTNAME_PATTERN      = /\A(?:[-\p{L}\d]{1,63}\.)+\p{L}{2,}\z/i.freeze
+  LOCAL_PART_PATTERN    = /\A[^@\s\."](?:\.?[^@\s\."])*\z/i.freeze
```
Comment (dkubb) [inline core/lib/spree/core/validators/email.rb]: Should we break out the pattern `[^@\s\."]` into something like `DISALLOWED_TOKEN` or similar and use it in both places within the regexp?

Comment (mbj) [inline core/lib/spree/core/validators/email.rb]: Good idea, but I'd name it `DISALLOWED_CHARACTERS` as we are not at the "token" level of understanding an email address.

File: core/lib/spree/core/validators/email.rb [hunk 4]
```diff
@@ -1,8 +1,23 @@
...
+  disallowed            = /[^@\s\."]/.freeze
+  LOCAL_PART_PATTERN    = /\A#{disallowed}(?:\.?#{disallowed})*\z/i.freeze
+
+  # From http://data.iana.org/TLD/tlds-alpha-by-domain.txt
+  TLDS =
+    IceNine.deep_freeze(
+      Spree::ROOT
+      .join(*%w[config tlds-alpha-by-domain.txt])
+      .read
+      .split
```
Comment (dkubb) [inline core/lib/spree/core/validators/email.rb]: I wonder if `Pathname#readlines` could be used to replace these two lines? I'm not sure if it chomps the line though.

Comment (mbj) [inline core/lib/spree/core/validators/email.rb]: No it does not, hence I used `read` and `split` as it spares a `map` over `chomp`.

**Pull request merged**